modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,358 @@
+
+----
+var idsList: Array<Number> = [1, 22, 333, 4444]
+----
+
+The variable `idsList` only passes the type-checking phase when the array that is assigned to it only contains values of the `Number` type.
+
+=== Object Type
+
+You can define an `Object` type in two ways: `Object` or `{}` both define an open object and don't specify any constraint to their key-value pairs. See also, <<close_open_objects>>.
+
+The syntax to define an `Object` type with a set of constraints for its key-value pairs is very similar to defining a value of an `Object` type:
+
+----
+{
+    keyName @(attrName: AttrType): valueType,
+    ...
+}
+----
+
+For example, to define a `User` type with `firstName` `String`, `lastName` `String `and `age` `Number`, the syntax is:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+type User = {
+    firstName: String,
+    lastName: String,
+    age: Number
+}
+----
+
+You can apply some modifiers to each key-value pair to determine if a field is always present or if the field can be repeated.
+
+For repeated fields use `&#42;`. For example, the following type allows you to have multiple `lastName` fields:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age: Number
+}
+----
+
+For conditional fields, use `?`. For example, the following type allows the age field to be present or not:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+[[close_open_objects]]
+==== Close and Open Objects
+
+A closed object means that the type only accepts an object value if there are no additional fields in the object,
+just those key-value pairs specified by the type.
+
+An open object only put constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accepts objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
+
+The following example works even if `age` is not mentioned in the type declaration because the type of the variable `user` is an open object:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
+----
+
+The syntax to specify a closed object is `{| |}`. The following example makes the `User` type support only the keys `firstName` and `lastName`. The script throws an exception because the `age` field is not accepted.
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var user: {|firstName: String, lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+
+=== Function Type
+
+DataWeave is a functional language, and functions are first class citizens, which means that functions are values,
```

Comment:
```suggestion
DataWeave is a functional language, and functions are considered first-class citizens, which means that functions are values with associated types
```
Is "first-class citizen" actually a technical term in this context?
How does having a value with an associated type make a function superior to something else (and what is that something else)?
Please check with writing team; I am not familiar with this concept.